### PR TITLE
[ISSUE #525] Support the message track

### PR DESCRIPTION
**Pull request by zongtanghu**
## What is the purpose of the change
Message track trace refers to the consumption processing of a message which is sent from the producer instance has arrived the broker,and then has been consumed by the consumer instance.In the whole process, the information of time, location and other related services(including Client and Broker) is aggregated into the complete link information of message communication.
In MQ system, the complete link of a message contains three roles: producer, broker server and consumer.In the process of message communicaton, each role adds relevant information to the track trace link.By aggregating these information,it can provide some powerful support to user.
About the Apache RocketMQ project,I will add the feature of message track trace which can help users query every complete link data of a message exactly.This funcation is important to RocketMQ,especially in some  financial application fields.
## Brief changelog
Look at the client/broker/common/tools/example/logappender/test module.
Look at the rocketmq/distribution/conf/2m-noslave
(1)broker-a.properties file.
(2)broker-b.properties file.
(3)broker-trace.properties file.
## Verifying this change
(1)the special storing for message track trace broker configuration files Join the flag as follows:
autoTraceBrokerEnable=true
(2)the normal broker configuration files Join the flag as follows:
autoTraceBrokerEnable=false

##Architecture and design
 We plan to design the function of message track trace including its store part and client instance collecting.The total architecture has two part below:
    （1）Client instance collects the infomation of message track trace 
    The design of client architecture above as follows:
    （a）."producer and consumer multi-thread module, and Blocking Queue"：here, in client,as a producer model,we can collect the infomation(such as,sending and consuming message cost time,broker store time,broker stored ip address and so on) and put this information into the Blocking Queue.And as s consumer model,use a thread called "MQ-AsyncArrayDispatcher" to take the message track trace infomation from Blocking Queue.Then this asynchronous thread( called "MQ-AsyncArrayDispatcher") pack the message track trace element as AsyncAppenderRequest task and submit to the thead pool.
    Last,the main execution process of  AsyncAppenderRequest task is sending the message track trace infomation which is collected above from client side to a special broker node which is redefined in below chapter.
    （b）.define a new pair hook which is implementation of the“SendMessageHook/ConsumeMessageHook”,from this,we can collect message track trace data  before and after publishing and subscribing messages.
    （2）Redefine a special broker node to store message track trace data
    As shown in the above picture,we can define a specail broker service node which can store the message track trace data in a common RocketMQ cluster.Here,we can add a flag(such as autoTraceBrokerEnable) in broker.properties file and use this variable to define Whether this broker is a specail node for storing message track trace data.
    （a）autoTraceBrokerEnable is false.This indicates this broker is an ordinary node,then "Trace_Topic" will not be created in this node.And it will still follow the original processing flow without any change.
    （b）autoTraceBrokerEnable is true.This indicates broker is an special node,which stores the message track trace data specailly.And The "Trace_Topic" is automatically created during the Broker's startup phase,this node automatically registers its owned set of topics in memory to nameserver(including Trace_Topic).Thus,in a RocketMQ cluster,there is only a specail broker node which stores message track trace datas.And clients(including publishing and subscribing messages) knows which broker node to send message trace data after they try fetch topic routing info from nameserver.
    （3）How to solve the query message track trace data by original topic                 
 For example,Topic which saves message track trace data is called "RMQ_SYS_TRACE_DATA_XXX" is instead of Topic of original data.But，message query(By messageId + topic,topic+ key or topic) which still uses original data on RocketMQ console side can't query the expected result for us.Here, we can fill in the keyset field of the message track trace data by using msgId (not offset MsgId) or Key of the original message content when send the message track trace data collected by the client, so the IndexFile index of the Broker end can be built according to the msgId or Key of the original message.
    And at the broker side,we can invokes the queryMessage() method of the storage part through the QueryMessageProcessor Business processor.
    （4）Finishing querying messages track trace datas in console
    Fisrtly,the console project is based on spring boot technology,so we can finish querying messages track track datas in this project.The basic design is using thread pool to send a RPC request to the redefined broker node to get the message track trace data.

Follow this checklist to help us incorporate your contribution quickly and easily. Notice, `it would be helpful if you could finish the following 5 checklist(the last one is not necessary)before request the community to review your PR`.

- [x] Make sure there is a [Github issue](https://github.com/apache/rocketmq/issues) filed for the change (usually before you start working on it). Trivial changes like typos do not require a Github issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue. 
- [x] Format the pull request title like `[ISSUE #123] Fix UnknownException when host config not exist`. Each commit in the pull request should have a meaningful subject line and body.
- [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [x] Write necessary unit-test(over 80% coverage) to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add integration-test in [test module](https://github.com/apache/rocketmq/tree/master/test).
- [x] Run `mvn -B clean apache-rat:check findbugs:findbugs checkstyle:checkstyle` to make sure basic checks pass. Run `mvn clean install -DskipITs` to make sure unit-test pass. Run `mvn clean test-compile failsafe:integration-test`  to make sure integration-test pass.
- [ ] If this contribution is large, please file an [Apache Individual Contributor License Agreement](http://www.apache.org/licenses/#clas).

### CODE REVIEW

File: client/src/main/java/org/apache/rocketmq/client/trace/core/dispatch/impl/AsyncArrayDispatcher.java [hunk 1]
```diff
@@ -0,0 +1,358 @@
...
+import java.util.HashSet;
+import java.util.ArrayList;
+import java.util.Map;
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicLong;
+
+/**
+ * Created by zongtanghu on 2018/11/6.
```
Comment (duhengforever) [inline client/src/main/java/org/apache/rocketmq/client/trace/core/dispatch/impl/AsyncArrayDispatcher.java]: Please remove author info

Comment (zongtanghu) [reply client/src/main/java/org/apache/rocketmq/client/trace/core/dispatch/impl/AsyncArrayDispatcher.java]: Okay,got it!Thank you very much.I will fix this issue.

Comment (vongosling) [reply client/src/main/java/org/apache/rocketmq/client/trace/core/dispatch/impl/AsyncArrayDispatcher.java]: @zongtanghu Thanks for your contribution, consider the big improvement for RocketMQ, I will pay some more time to help to review and merge it.

Comment (zongtanghu) [reply client/src/main/java/org/apache/rocketmq/client/trace/core/dispatch/impl/AsyncArrayDispatcher.java]: You are welcome！It's my honor to be able to join in the rocketmq community and contribute some features and documents.I'm very glad to contribute much more important features to RocketMQ 5.0 in the future,such as message routing and monitoring alarm.

Comment (zongtanghu) [reply client/src/main/java/org/apache/rocketmq/client/trace/core/dispatch/impl/AsyncArrayDispatcher.java]: Hi @vongosling @duhengforever @dongeforever ,I committed and pushed the rocketmq console part's codes for message track trace function.This is the pr link:https://github.com/apache/rocketmq-externals/pull/154.

**Comment by zongtanghu** [comment]
rocketmq client

File: client/src/main/java/org/apache/rocketmq/client/trace/core/common/TrackTraceType.java [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.rocketmq.client.trace.core.common;
+
+/**
+ * Created by zongtanghu on 2018/11/6.
+ */
```
[the remark this answers is not in the document]
Comment (zongtanghu) [reply client/src/main/java/org/apache/rocketmq/client/trace/core/common/TrackTraceType.java]: Okay,I will fix it in a new commit.

File: client/src/main/java/org/apache/rocketmq/client/consumer/DefaultMQPushConsumer.java [hunk 3]
```diff
@@ -261,10 +276,29 @@ public DefaultMQPushConsumer() {
      * @param allocateMessageQueueStrategy message queue allocating algorithm.
      */
     public DefaultMQPushConsumer(final String consumerGroup, RPCHook rpcHook,
-        AllocateMessageQueueStrategy allocateMessageQueueStrategy) {
+        AllocateMessageQueueStrategy allocateMessageQueueStrategy, boolean msgTraceSwitch) {
         this.consumerGroup = consumerGroup;
         this.allocateMessageQueueStrategy = allocateMessageQueueStrategy;
         defaultMQPushConsumerImpl = new DefaultMQPushConsumerImpl(this, rpcHook);
+        //if client open the message track trace feature
```
[the remark this answers is not in the document]
Comment (zongtanghu) [reply client/src/main/java/org/apache/rocketmq/client/consumer/DefaultMQPushConsumer.java]: Okay,I will fix it in a new commit.

File: broker/src/main/java/org/apache/rocketmq/broker/processor/AdminBrokerProcessor.java [hunk 4]
```diff
@@ -228,6 +228,14 @@ private synchronized RemotingCommand updateAndCreateTopic(ChannelHandlerContext
             return response;
         }
 
+        if (this.brokerController.getBrokerConfig().isAutoTraceBrokerEnable()) {
```
[the remark this answers is not in the document]
Comment (zongtanghu) [reply broker/src/main/java/org/apache/rocketmq/broker/processor/AdminBrokerProcessor.java]: Thanks for your advice.I will fix this code in a new commit.

**Comment by xiellggl** [comment]
您好我想问下两个问题
1.在broker.properties 里面加了autoTraceBrokerEnable=true然后broker启动后没显示这个配置项，版本是4.4.0。
2.在RocketMq-console-ng 中的 messagetrace 中按id能查到对应消息，点击MESSAGE TRACE DETAIL时显示：query message by key finished, but no message。
希望您有空能帮我解答一下

**Pull request merged**